lib/platform/linux-uart.c

```diff
@@ -0,0 +1,441 @@
+/*
+ * Microsemi Switchtec(tm) PCIe Management Library
+ * Copyright (c) 2017, Microsemi Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#ifdef __linux__
+
+#include "../switchtec_priv.h"
+#include "switchtec/switchtec.h"
+#include "gasops.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stddef.h>
+#include <assert.h>
+#include <string.h>
+
+#include <termios.h>
+#include "../crc8.h"
+
+const char *dbginfo_off = "pscdbg 0 all\r";
+const char *echo_off = "echo 0\r";
+const char *echo_on = "echo 1\r";
+
+struct switchtec_uart{
+	struct switchtec_dev dev;
+	int fd;
+};
+
+#define to_switchtec_uart(d) \
+	((struct switchtec_uart *) \
+	 ((char *)d - offsetof(struct switchtec_uart, dev)))
+
+
+#define UART_MAX_WRITE_BYTES			(100)
+#define RETRY_NUM				(50)
+
+static int cli_control(struct switchtec_dev *dev, const char *str)
+{
+	int ret;
+	int idx;
+	int rty_cnt = 0;
+	char rtn[1024];
+	struct switchtec_uart *udev = to_switchtec_uart(dev);
+
+	ret = tcflush(udev->fd,TCIOFLUSH);
```

Comment:
tcflush used to flushing both the out and in buffer, with TCIOFLUSH flag set.
yes, there maybe chars between flush and write, but the difficult is the chars from FW is a superset, utility code could not process all the cases.
Here use the tcflush, is mainly for flush the ">" symbol if exist, so the following check of the ">" symbol, if it appear, we consider, the request's respose is finished.

Reply:
Yes, as I said, you are much better reading that data back than flushing. It's much better if you parse it all. It kind of looks like there's a counter in the prompt and if you can use it to ensure you get right prompt for the command you sent that would be much better.

Reply:
We just care about the end of the prompt, instead of the ensure the counter, because
1. if the counter return from FW is correct, but for the reaon of the link is bad the counter is corrpted, this will casue unnecessary retry
2. CRC only protect the offset and payload. 
to take counter sequence into account, it make the thing heavy, but not necessary
3. To track the counter, a global variable is introduce, because the counter is  for all the commands: cli_control, gasrd, gaswr.